syntax = "proto2";

package rv.protos;

import "rastervision/protos/machine_learning.proto";
import "rastervision/protos/project.proto";

message ProcessTrainingDataConfig {
    message ObjectDetectionOptions {
        /*
            The ratio of negative chips (those containing no bounding boxes) to
            positive chips. This can be useful if the statistics of the
            background is different in positive chips. For example, in car
            detection, the positive chips will always contain roads, but no
            examples of rooftops since cars tend to not be near rooftops.
        */
        required float neg_ratio = 1;

        /*
            When a box is partially outside of a training chip, it is not
            clear if (a clipped version) of the box should be included in
            the chip. If the IOA (intersection over area) of the box with the
            chip is greater than ioa_thresh, it is included in the chip.
        */
        optional float ioa_thresh = 2 [default=0.8];

        /*
            Different models in the Object Detection API have different inputs.
            Some models allow variable size inputs so several methods
            of building training data are required

            Valid values are:
                - chip (default)
                - label
                    - each label's bounding box is the positive window
                - image
                    - each image is the positive window
        */
        optional string window_method = 3 [default="chip"];

        /*
            If method is "label", the positive window can be buffered.
            If value is >= 0. and < 1., the value is treated as a percentage
            If value is >= 1., the value is treated in number of pixels
        */
        optional float label_buffer = 4 [default=0.];

    }

    message ClassificationOptions {

    }

    message Options {
        optional int32 chip_size = 1;
        optional bool debug = 2 [default=true];

        // The directory where all the output should be placed.
        optional string output_uri = 3;

        oneof ml_options_type {
            ObjectDetectionOptions object_detection_options = 4;
            ClassificationOptions classification_options = 5;
        }
    }

    repeated Project train_projects = 1;
    repeated Project validation_projects = 2;
    required MachineLearning machine_learning = 3;
    required Options options = 4;
}
